universal-login-commons/lib/core/utils/securityCodes.ts

```diff
@@ -32,6 +29,31 @@ export const isValidCode = (code: number[], publicKey: string) => {
   return deepEqual(expectedCode, code);
 };
 
+export const findValidAddressFromCodePart = (enteredCode: number[], incomingPublicKeys: string[]) => {
+  const possibleAddresses = findPossibleAddressesFromCodePart(enteredCode, incomingPublicKeys);
+  return possibleAddresses.length === 1 ? possibleAddresses[0] : undefined;
+};
+
+export const findPossibleAddressesFromCodePart = (enteredCode: number[], incomingPublicKeys: string[]) => {
+  const generatedCodes = incomingPublicKeys.map((key) => generateCode(key));
+  const validIndicies = findIndiciesOfValidCodesByPrefix(generatedCodes, enteredCode);
+  return getArrayElementsFromIndicies(incomingPublicKeys, validIndicies);
+};
+
+export const findIndiciesOfValidCodesByPrefix = (codes: number[][], enteredCode: number[]) =>
+  codes.reduce((indices, code, index) => {
```

Comment:
If you want to do reduce on arrays, you probably should do map and filter instead

Reply:
Did you mean?
```ts
functoin filterAddressesByPrefix(prefix: [], addresses : [][]) {
  addresses
    .map((address, index) => [address, index])
    .filter(([address, index]) => startsWith(prefix, address))
    .map(([, index]) => index);
}
```

Reply:
Yeah, I meant that ;)